js/src/components/tree-select-control/index.js

```diff
@@ -1,56 +1,190 @@
 /**
  * External dependencies
  */
-import { useState } from '@wordpress/element';
+import { useMemo, useState } from '@wordpress/element';
 import classnames from 'classnames';
-// import { __, _n, sprintf } from '@wordpress/i18n';
-// import { Component, createElement } from '@wordpress/element';
-// import { debounce, escapeRegExp, identity, noop } from 'lodash';
-// import PropTypes from 'prop-types';
-// import { withFocusOutside, withSpokenMessages } from '@wordpress/components';
-// import { withInstanceId, compose } from '@wordpress/compose';
+// eslint-disable-next-line import/no-extraneous-dependencies,@woocommerce/dependency-group,@wordpress/no-unsafe-wp-apis
+import { __experimentalUseFocusOutside as useFocusOutside } from '@wordpress/compose';
```

Comment:
Thanks for opening the discussion

>  How did you know about this? 😆

If you check the [Select Control Component ](https://github.com/woocommerce/woocommerce-admin/blob/main/packages/components/src/select-control/index.js#L9) which is the base model for Tree Select Control (as dicsussed), you can see how they use the `withFocusOutside` which is the HoC version. 

Since we are committed into building a functional version we need to rely on the hook version `useFocusOutside` which is still experimental (as other components like `Text` in our codebase). I don't think this is a huge problem if we are careful in testing.

It can also help to promote the hook to NON-Experimental or even fix some bugs if we found them. In the end, the idea is also to contribute to Woo/WP in terms of adding a new component (TreeSelectControl) that will be also experimental at the beginning. So, except if something is going wrong with it, I think we can give it a chance. We have always time to implement our custom solution if it doesn't work.

> Also, I notice we don't have @wordpress/compose specified in package.json, so if we really want to use this, we should add @wordpress/compose into package.json. Note that it will be bundled by webpack

Thanks! Noted. Still having troubles with the full understanding of dependencies, "auto" injected dependencies etc... 👌 

> Alternatively, we could also consider copying the [useFocusOutside code](https://github.com/WordPress/gutenberg/blob/7056cc19240e47a0282edb3b820574f09b9eb680/packages/compose/src/hooks/use-focus-outside/index.js) into our project here.

What's the advantage of this against the usage of the WP code? 🤔

Reply:
> What's the advantage of this against the usage of the WP code? 🤔

If we use the `__experimentalUseFocusOutside` from the package, it may change and break our things without us knowing. By copying into our project, then at least it is controlled by us, but then maintaining it / keeping it up to date would be another issue. This topic is somewhat arguable, I guess since it is working fine for us, we can move with what we have here.